src/newscale/stages.py

```diff
@@ -1,365 +1,497 @@
-        self.sock.sendall(cmd_bytes)
-        resp = self.sock.recv(1024).decode('utf-8').strip('<>\r')
-        SSSSSS = int(resp.split()[1], 16)
-        PPPPPPPP = int(resp.split()[2], 16)
-        EEEEEEEE = int(resp.split()[3], 16)
-        return SSSSSS, PPPPPPPP, EEEEEEEE
-        
-    # 19
-    def readMotorFlags(self):
-        """
-        This command reports internal flags used by the controller to monitor
-        motor conditions.
-        """
-        cmd = "<19>\r"
-        cmd_bytes = bytes(cmd, 'utf-8')
-        self.sock.sendall(cmd_bytes)
-        resp = self.sock.recv(1024).decode('utf-8').strip('<>\r')
-        # TODO parse response
+        return self._send(self._get_cmd_str(Cmd.HALT))
+
+    # <04>
+    def run(self, direction: Direction, seconds: float = None):
+        """Enable motor for a specified time or without limit if unspecified.
+
+        :param direction: forward or reverse
+        :param seconds: time to run the motor in seconds (with increments of
+            tenths of seconds).
+        """
+        assert direction is not Direction.NEITHER, \
+            "Direction must be forward or reverse."
+        if seconds is None:
+            return self._send(self._get_cmd_str(Cmd.RUN, direction.value))
+        duration = round(seconds*10)  # value is encoded in tenths of seconds.
+        assert duration < 256, "Time input value exceeds maximum value."
+        return self._send(self._get_cmd_str(Cmd.RUN, direction.value,
+                                            f"{duration:04x}"))
+
+    # <05>
+    #def time_step(self, direction: Direction, step_count: int = None,
+    #              step_interval_us: float = None,
+    #              step_duration_us: float = None):
+    #    """Setup a specified number of steps to move through at a specific
+    #    interval."""
+    #    # Note: we cannot accept NEITHER as a direction
+    #    raise NotImplementedError
+
+    # <06>
+    def distance_step(self, direction: Direction, step_size_mm: float = None):
+        """Take a step of size `step_size` in the specified direction.
+        If no step size is specified, the previous step size will be used.
+
+        :param direction: direction to step in.
+        :param step_size_mm: the size of the step. Optional.
+        """
+        # TODO: warn if step size was never specified.
+        step_size_ticks = round(step_size_mm*TICKS_PER_MM)  # 32 bit unsigned.
+        assert step_size_ticks.bit_length() < 32, "Step size exceeds maximum."
+        return self._send(self._get_cmd_str(Cmd.STEP_CLOSED_LOOP,
+                                            direction.value,
+                                            f"{step_size_ticks:08x}"))
+
+    # <06> variant
+    def set_distance_step_size(self, step_size_mm: float):
+        """Specify the step size taken (in mm) in :meth:`distance_step`"""
+        return self.distance_step(Direction.NEITHER, step_size_mm)
+
+    # <07>
```

Comment:
please implement this

Reply:
I skipped this since you already get the same behavior from <20> (View and select open-or-closed-loop mode) with more specific intent. Issuing <07> toggles between open/closed-loop modes without telling you what mode it's on.

I think exposing this could honestly lead to more confusion to the user ("what mode am I on now?") while <20> already covers this and does a better job.

Reply:
If you really *really* want me to expose it, I can.

Reply:
actually <07> has to do with absolute/relative mode, which is different than open/closed-loop mode. the command reference i sent you lists it as "Toggle Absolute/Relative Position", but the (newer?) one you included calls it "Clear the encoder count". at any rate, it has something to do with relative encoding and i think we should implement it.

Reply:
implemented. According to the Newscale Pathway Software datasheet, this command zeros the encoder count, which has the same effect of setting the current position to zero. (Now it makes sense why you can set software limits to be negative numbers.)